src/mono/wasm/runtime/diagnostics.ts

```diff
@@ -0,0 +1,142 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { Module } from "./imports";
+import cwraps from "./cwraps";
+import type { EventPipeSessionOptions } from "./types";
+import type { VoidPtr } from "./types/emscripten";
+import * as memory from "./memory";
+
+const sizeOfInt32 = 4;
+
+export type EventPipeSessionID = bigint;
+type EventPipeSessionIDImpl = number;
+
+/// An EventPipe session object represents a single diagnostic tracing session that is collecting
+/// events from the runtime and managed libraries.  There may be multiple active sessions at the same time.
+/// Each session subscribes to a number of providers and will collect events from the time that start() is called, until stop() is called.
+/// Upon completion the session saves the events to a file on the VFS.
+/// The data can then be retrieved as Blob or as a data URI (prefer Blob).
+export interface EventPipeSession {
+    // session ID for debugging logging only
+    get sessionID(): EventPipeSessionID;
+    start(): void;
+    stop(): void;
+    getTraceBlob(): Blob;
+}
+
+// internal session state of the JS instance
+enum State {
+    Initialized,
+    Started,
+    Done,
+}
+
+function withStackAlloc<TRes> (bytesWanted: number, f : (ptr: VoidPtr) => TRes): TRes {
+    const sp = Module.stackSave();
+    const ptr = Module.stackAlloc (bytesWanted);
+    try {
+        return f (ptr);
+    } finally {
+        Module.stackRestore(sp);
+    }
+}
+
+function start_streaming (sessionID: EventPipeSessionIDImpl): void {
+    cwraps.mono_wasm_event_pipe_session_start_streaming (sessionID);
+}
+
+function stop_streaming (sessionID: EventPipeSessionIDImpl): void {
+    cwraps.mono_wasm_event_pipe_session_disable (sessionID);
+}
+
+function makeTimestamp(): string
+{
+    // ISO date string, but with : and . replaced by -
+    const t = new Date();
+    const s = t.toISOString();
+    return s.replace(/[:.]/g, "-");
+}
+
+// The name of the session trace file on the VFS.  This is an implementation detail that is not exposed to the user.
+function computeTracePath (): string {
```

Comment:
nit: could be inlined.

Reply:
I decided to just use consecutive numbering for the VFS files.  the timestamp wasn't effective anyway if sessions are created below the timestamp resolution.